httpx/streams.py

```diff
@@ -0,0 +1,297 @@
+import mimetypes
+import os
+import typing
+from io import BytesIO
+from json import dumps as json_dumps
+from pathlib import Path
+from urllib.parse import urlencode
+
+from .utils import format_form_param
+
+RequestData = typing.Union[dict, str, bytes, typing.AsyncIterator[bytes]]
+
+RequestFiles = typing.Dict[
+    str,
+    typing.Union[
+        # file (or str)
+        typing.Union[typing.IO[typing.AnyStr], typing.AnyStr],
+        # (filename, file (or str))
+        typing.Tuple[
+            typing.Optional[str], typing.Union[typing.IO[typing.AnyStr], typing.AnyStr],
+        ],
+        # (filename, file (or str), content_type)
+        typing.Tuple[
+            typing.Optional[str],
+            typing.Union[typing.IO[typing.AnyStr], typing.AnyStr],
+            typing.Optional[str],
+        ],
+    ],
+]
+
+
+class Stream:
+    async def __aiter__(self) -> typing.AsyncIterator[bytes]:
+        yield b""
+
+    async def aclose(self) -> None:
+        pass
+
+    async def aread(self) -> bytes:
+        return b"".join([part async for part in self])
+
+
+class ResponseStream(Stream):
+    def __init__(self, iterator: typing.AsyncIterator[bytes], close: typing.Callable):
+        self.iterator = iterator
+        self.close_func = close
+
+    async def __aiter__(self) -> typing.AsyncIterator[bytes]:
+        async for chunk in self.iterator:
+            yield chunk
+
+    async def aclose(self) -> None:
+        await self.close_func()
+
+
+class RequestStream(Stream):
+    """
+    Base class for streaming request content.
+    Defaults to a "no request body" implementation.
+    """
+
+    def get_headers(self) -> typing.Dict[str, str]:
+        """
+        Return a dictionary of request headers that are implied by the encoding.
+        """
+        return {}
+
+    def can_replay(self) -> bool:
+        """
+        Return `True` if `__aiter__` can be called multiple times.
+
+        We need this in order to determine if we can re-issue a request body
+        when we receive a redirect response.
+        """
+        return True
+
+
+class BytesRequestStream(RequestStream):
+    """
+    Request content encoded as plain bytes.
+    """
+
+    def __init__(self, body: typing.Union[str, bytes]) -> None:
+        self.body = body.encode("utf-8") if isinstance(body, str) else body
+
+    def get_headers(self) -> typing.Dict[str, str]:
+        content_length = str(len(self.body))
+        return {"Content-Length": content_length}
+
+    async def __aiter__(self) -> typing.AsyncIterator[bytes]:
+        yield self.body
+
+
+class IteratorRequestStream(RequestStream):
+    """
+    Request content encoded as plain bytes, using an async byte iterator.
+    """
+
+    def __init__(self, aiterator: typing.AsyncIterator[bytes]) -> None:
+        self.aiterator = aiterator
+
+    def can_replay(self) -> bool:
+        return False
+
+    def get_headers(self) -> typing.Dict[str, str]:
+        return {"Transfer-Encoding": "chunked"}
```

Comment:
Is `Transfer-Encoding` actually a valid/actionable request header? The [MDN web docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding) present it as a response-only header. 🤔 

Even then, we're not making sure that the parts in the `aiterator` are separated with `\r\n`, are we?

(Not directly related to this PR, we probably were doing this before.)

Reply:
Not sure why they list it as response only, since it's required for streaming unknown-length request bodies too. Wrt. the framing, `h11` has us covered... https://h11.readthedocs.io/en/latest/api.html#message-body-framing-content-length-and-all-that